<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Zachary S. L. Foster and Niklaus J. Grünwald" />

<meta name="date" content="2016-08-20" />

<title>An R package for metabarcoding research</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
  padding-left: 10px;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MetacodeR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="vignettes--01--extracting_taxonomy_data.html">Parsing input files</a>
    </li>
    <li>
      <a href="vignettes--02--plotting_taxonomic_data.html">Heat tree plotting</a>
    </li>
    <li>
      <a href="vignettes--03--manipulating_taxonomic_data.html">Subsetting</a>
    </li>
    <li>
      <a href="vignettes--04--hierarchical_subsampling.html">Sampling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Publication Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="publication--05--16s_human_microbiome.html">Human microbiome</a>
    </li>
    <li>
      <a href="publication--06--tara_oceans.html">TARA Oceans</a>
    </li>
    <li>
      <a href="publication--01--silva.html">In silico PCR: SILVA</a>
    </li>
    <li>
      <a href="publication--02--rdp.html">In silico PCR: RDP</a>
    </li>
    <li>
      <a href="publication--03--greengenes.html">In silico PCR: Greengenes</a>
    </li>
    <li>
      <a href="publication--04--16s_database_comparision.html">In silico PCR: Comparing databases</a>
    </li>
  </ul>
</li>
<li>
  <a href="presentation_page.html">Presentation</a>
</li>
<li>
  <a href="https://github.com/grunwaldlab/metacoder">GitHub Repository</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">An R package for metabarcoding research</h1>
<h4 class="author"><em>Zachary S. L. Foster and Niklaus J. Grünwald</em></h4>
<h4 class="date"><em>2016-08-20</em></h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Metabarcoding is revolutionizing microbial ecology and presenting new challenges:</p>
<ul>
<li>Numerous database formats make taxonomic data difficult to parse, combine, and subset.</li>
<li>Stacked bar charts, commonly used to depict community diversity, lack taxonomic context.</li>
<li>Barcode loci and primers are a source of under-explored bias.</li>
</ul>
<p>MetacodeR is an R package that attempts to addresses these issues:</p>
<ul>
<li>Sources of taxonomic data can be extracted from any file format and manipulated.</li>
<li>Community diversity can be visualized by color and size in a tree plot.</li>
<li>Primer specificity can be estimated with <em>in silico</em> PCR.</li>
</ul>
</div>
<div id="documentation" class="section level2">
<h2>Documentation</h2>
<p>Documentation is under construction at <a href="http://grunwaldlab.github.io/metacoder" class="uri">http://grunwaldlab.github.io/metacoder</a>.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>While this project is in development it can be installed through Github:</p>
<p>devtools::install_github(repo=“grunwaldlab/metacoder”, build_vignettes = TRUE) library(metacoder)</p>
<p>If you’ve built the vignettes, you can browse them with:</p>
<p>browseVignettes(package=“metacoder”)</p>
</div>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>The function that runs <em>in silico</em> PCR requires <code>primersearch</code> from the EMBOSS (Olson, 2002) tool kit to be installed. This is not an R package, so it is not automatically installed. Type <code>?primersearch</code> after installing and loading MetcodeR for installation instructions.</p>
</div>
<div id="extracting-taxonomic-data" class="section level2">
<h2>Extracting taxonomic data</h2>
<p>Most databases have a unique file format and taxonomic hierarchy/nomenclature. Taxonomic data can be extracted from any file format using the <strong>extract_taxonomy</strong> function. Classifications can be parsed offline or retrieved from online databases if a taxon name, taxon ID, or sequence ID is present. A regular expression with capture groups and a corresponding key is used to define how to parse the file. The example code below parses the 16s Ribosome Database Project training set for Mothur.</p>
<p>R can be used to download files from the internet and decompress them. The code below downloads the compressed data to a temporary directory:</p>
<pre class="r"><code>rdp_fasta_url &lt;- &quot;http://mothur.org/w/images/b/b5/Trainset10_082014.rdp.tgz&quot;
temp_dir_path &lt;- tempdir()
local_file_path &lt;- file.path(temp_dir_path, basename(rdp_fasta_url))
download.file(url = rdp_fasta_url, destfile = local_file_path, quiet = TRUE)</code></pre>
<p>Next we will uncompress the archive and identify the fasta file.</p>
<pre class="r"><code># Get contents of tar archive
unpacked_file_paths &lt;- untar(local_file_path, list = TRUE)
# Uncompress archive
untar(local_file_path, exdir = temp_dir_path)
# Identify the Mothur RDP training set
unpacked_fasta_path &lt;- file.path(temp_dir_path, 
                                 unpacked_file_paths[grepl(&quot;fasta$&quot;, unpacked_file_paths)])</code></pre>
<p>The file can then be parsed using the <code>ape</code> package and the taxonomy data in the headers can be extracted by <code>extract_taxonomy</code>:</p>
<pre class="r"><code># Load the package
library(metacoder)
# Load the input FASTA file
seqs &lt;- ape::read.FASTA(unpacked_fasta_path)
# Print an example of the sequence headers
cat(names(seqs)[1])</code></pre>
<pre><code>## AB294171_S001198039  Root;Bacteria;Firmicutes;Bacilli;Lactobacillales;Carnobacteriaceae;Alkalibacterium</code></pre>
<pre class="r"><code># Extract the taxonomic information of the sequences
data &lt;- extract_taxonomy(seqs, regex = &quot;^(.*)\\t(.*)&quot;,
                         key = c(id = &quot;obs_info&quot;, &quot;class&quot;),
                         class_sep = &quot;;&quot;)</code></pre>
<p>The resulting object contains sequence information associated with an inferred taxonomic hierarchy. The standard print method of the object shows a part of every kind of data it contains:</p>
<pre class="r"><code>data</code></pre>
<pre><code>## `taxmap` object with data for 2794 taxa and 10650 observations:
## 
## ---------------------------- taxa ----------------------------
## 1, 2, 3, 4, 5 ... 2789, 2790, 2791, 2792, 2793, 2794
## 
## ------------------------- taxon_data -------------------------
## # A tibble: 2,794 x 3
##   taxon_ids supertaxon_ids            name
##       &lt;chr&gt;          &lt;chr&gt;           &lt;chr&gt;
## 1         1           &lt;NA&gt;            Root
## 2         2              1         Archaea
## 3         3              1        Bacteria
## 4         4              2 &quot;Crenarchaeota&quot;
## 5         5              2 &quot;Euryarchaeota&quot;
## 6         6              2  &quot;Korarchaeota&quot;
## 7         7              2 &quot;Nanoarchaeota&quot;
## # ... with 2,787 more rows
## 
## -------------------------- obs_data --------------------------
## # A tibble: 10,650 x 3
##   obs_taxon_ids                  id
##           &lt;chr&gt;               &lt;chr&gt;
## 1          1289 AB294171_S001198039
## 2           369 AB243007_S000622964
## 3          2402 AJ717394_S000623308
## 4          1409 AJ518871_S000351779
## 5           962 CP000159_S000632736
## 6          1311   X76329_S000015720
## 7           616 AB184711_S000652435
## # ... with 1.064e+04 more rows, and 1 more variables: sequence &lt;chr&gt;
## 
## ------------------------- taxon_funcs -------------------------
## n_obs, n_obs_1 ... n_subtaxa_1, hierarchies</code></pre>
<p>Note the <code>taxon_funcs</code> part contains function names (e.g. <code>n_obs</code>) used to calculate taxon statistics. You can see the results of these calculations using the <code>taxon_data</code> function:</p>
<pre class="r"><code>taxon_data(data)</code></pre>
<pre><code>## # A tibble: 2,794 x 9
##    taxon_ids supertaxon_ids              name n_obs n_obs_1 n_supertaxa
##        &lt;chr&gt;          &lt;chr&gt;             &lt;chr&gt; &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;
## 1          1           &lt;NA&gt;              Root 10650       0           0
## 2          2              1           Archaea   410       0           1
## 3          3              1          Bacteria 10240       0           1
## 4          4              2   &quot;Crenarchaeota&quot;    55       0           2
## 5          5              2   &quot;Euryarchaeota&quot;   335       0           2
## 6          6              2    &quot;Korarchaeota&quot;    10      10           2
## 7          7              2   &quot;Nanoarchaeota&quot;     3       3           2
## 8          8              2 Nanohaloarchaeota     1       0           2
## 9          9              2  &quot;Thaumarchaeota&quot;     6       0           2
## 10        10              4      Thermoprotei    55       0           3
## # ... with 2,784 more rows, and 3 more variables: n_subtaxa &lt;dbl&gt;,
## #   n_subtaxa_1 &lt;dbl&gt;, hierarchies &lt;chr&gt;</code></pre>
<p>There is corresponding <code>obs_data</code> function and <code>obs_funcs</code> list, but they are not used in this case.</p>
</div>
<div id="heat-trees" class="section level2">
<h2>Heat trees</h2>
<p>The hierarchical nature of taxonomic data makes it difficult to plot effectively. Most often, bar charts, stacked bar charts, or pie graphs are used, but these are ineffective when plotting many taxa or multiple ranks. MetacodeR maps taxonomic data (e.g. sequence abundance) to color/size of tree components in what we call a <strong>Metadiversity Plot</strong>:</p>
<pre class="r"><code>heat_tree(data, node_size = n_obs, node_label = name, node_color = n_obs)</code></pre>
<p><img src="index_files/figure-html/home_plot_1-1.png" width="672" /></p>
<p>The default size range displayed is optimized for each plot. The legend represents the number of sequences for each taxon as both a color gradient and width of nodes. Only a few options are needed to make effective plots, yet many are available for customization of publication-ready graphics:</p>
<pre class="r"><code>set.seed(8)
heat_tree(data, node_size = n_obs, edge_color = n_supertaxa,
          node_label = name, node_color = n_obs,
          node_color_range = c(&quot;cyan&quot;, &quot;magenta&quot;, &quot;green&quot;),
          edge_color_range   = c(&quot;#555555&quot;, &quot;#EEEEEE&quot;),
          initial_layout = &quot;reingold-tilford&quot;, layout = &quot;davidson-harel&quot;,
          overlap_avoidance = 0.5)</code></pre>
<pre><code>## Error in UseMethod(&quot;depth&quot;) : 
##   no applicable method for &#39;depth&#39; applied to an object of class &quot;NULL&quot;
## Error in UseMethod(&quot;depth&quot;) : 
##   no applicable method for &#39;depth&#39; applied to an object of class &quot;NULL&quot;
## Error in UseMethod(&quot;depth&quot;) : 
##   no applicable method for &#39;depth&#39; applied to an object of class &quot;NULL&quot;
## Error in grid.Call.graphics(L_upviewport, as.integer(n)) : 
##   cannot pop the top-level viewport (&#39;grid&#39; and &#39;graphics&#39; output mixed?)
## Error in grid.Call.graphics(L_upviewport, as.integer(n)) : 
##   cannot pop the top-level viewport (&#39;grid&#39; and &#39;graphics&#39; output mixed?)</code></pre>
<p>The above command can take several minutes since it uses a force-directed layout that requires simulations.</p>
<p>Note that <code>plot</code> is a generic R function that works differently depending on what it is given to it. MetacodeR supplies the function <strong>plot_taxonomy</strong>, which is used when plot is given the type of data outputted by <code>extract_taxonomy</code>. To see the long list of available plotting options, type <code>?plot_taxonomy</code>.</p>
</div>
<div id="subsetting" class="section level2">
<h2>Subsetting</h2>
<p>Taxonomic data in the form used in metacodeR can be manipulated using functions inspired by <code>dplyr</code>. For example, taxa can be subset using <code>filter_taxa</code>. Unlike the <code>dplyr</code> function <code>filter</code>, users can choose preserve or remove the subtaxa, supertaxa, and associated observation data of the selected taxa. For example, <code>filter_taxa</code> can be used to look at just the Archaea:</p>
<pre class="r"><code>set.seed(1)
heat_tree(filter_taxa(data, name == &quot;Archaea&quot;, subtaxa = TRUE),
          node_size = n_obs, node_label = name, 
          node_color = n_obs, layout = &quot;fruchterman-reingold&quot;)</code></pre>
<p><img src="index_files/figure-html/home_plot_3-1.png" width="672" /></p>
<p>Any column displayed by <code>taxon_data</code> can be used with <code>filter_taxa</code> (and most other metacodeR commands) as if it were a variable on its own. To make the Archaea-Bacteria division more clear, the “Root” taxon can be removed, resulting in two separate trees:</p>
<pre class="r"><code>subsetted &lt;- filter_taxa(data, n_supertaxa &gt; 0)
set.seed(2)
heat_tree(subsetted, node_size = n_obs, node_label = name,
          node_color = n_obs, tree_label = name)</code></pre>
<p><img src="index_files/figure-html/home_plot_4-1.png" width="672" /></p>
<p>Although observations (information in <code>obs_data</code>) are typically assinged to the tips of the taxonomy, they can also be assigned to any internal node. When a taxon is removed by a filtering operation, the observations assinged to it are reassigned to an unfiltered supertaxon by default. This makes it easy to remove lower ranks of a taxonomy without discarding oberservations assinged to the tips:</p>
<pre class="r"><code>set.seed(1)
filter_taxa(data, n_supertaxa &lt; 4) %&gt;%
  heat_tree(node_size = n_obs, node_label = name, node_color = n_obs)</code></pre>
<p><img src="index_files/figure-html/home_plot_5-1.png" width="672" /></p>
<p>There is also a <code>filter_obs</code> function which can filter out observations and the taxa they are assigned to.</p>
</div>
<div id="sampling" class="section level2">
<h2>Sampling</h2>
<p>When calculating statistics for taxa, the amount of data should be balanced across taxa and there should be enough data per taxon to provide unbiased estimates. Random samples from large reference databases are biased towards overrepresented taxa. MetacodeR provides two ways to randomly sample taxonomic data. The function <code>taxonomic_sample</code> is used to create taxonomically balanced random samples. The acceptable range of sequence or subtaxa counts can be defined for each taxonomic rank; taxa with too few are excluded and taxa with too many are randomly subsampled. The code below samples the data such that rank 6 taxa will have 5 sequences and rank 3 taxa (phyla) will have less than 100:</p>
<pre class="r"><code>set.seed(1)
sampled &lt;- taxonomic_sample(subsetted, max_counts = c(&quot;3&quot; = 100, &quot;6&quot; = 5), min_counts = c(&quot;6&quot; = 5))
sampled &lt;- filter_taxa(sampled, n_obs &gt; 0, subtaxa = FALSE) </code></pre>
<pre class="r"><code>set.seed(3)
heat_tree(sampled, node_size = n_obs, node_label = n_obs, overlap_avoidance = 0.5,
          node_color = n_obs, tree_label = name)</code></pre>
<p><img src="index_files/figure-html/home_plot_6-1.png" width="672" /></p>
<p>This can also be accomplished using the <code>dplyr</code> equivalents to <code>sample_n</code> and <code>sample_frac</code> by weighting the probability of sampling observations by the inverse of the number of observations:</p>
<pre class="r"><code>set.seed(6)
sample_n_obs(subsetted, size = 1000, taxon_weight = 1 / n_obs, unobserved = FALSE) %&gt;%
  heat_tree(node_size = n_obs, node_label = n_obs, overlap_avoidance = 0.5,
            node_color = n_obs, tree_label = name)</code></pre>
<p><img src="index_files/figure-html/home_plot_7-1.png" width="672" /></p>
</div>
<div id="in-silico-pcr" class="section level2">
<h2>In silico PCR</h2>
<p>The function <strong>primersearch</strong> is a wrapper for an EMBOSS tool that implements <em>in silico</em> PCR. The code below estimates the coverage of the universal prokaryotic primer pair 515F/806R (Walters Hyde, et al., 2015):</p>
<pre class="r"><code>pcr &lt;- primersearch(sampled, 
                    forward = c(&quot;515F&quot; = &quot;GTGYCAGCMGCCGCGGTAA&quot;), 
                    reverse = c(&quot;806R&quot; = &quot;GGACTACNVGGGTWTCTAAT&quot;), 
                    mismatch = 10)
taxon_data(pcr)</code></pre>
<pre><code>## # A tibble: 1,999 x 11
##    taxon_ids supertaxon_ids              name n_obs n_obs_1 n_supertaxa
##        &lt;chr&gt;          &lt;chr&gt;             &lt;chr&gt; &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;
## 1          2           &lt;NA&gt;           Archaea   362       0           0
## 2          3           &lt;NA&gt;          Bacteria  4569       0           0
## 3          4              2   &quot;Crenarchaeota&quot;    55       0           1
## 4          5              2   &quot;Euryarchaeota&quot;   287       0           1
## 5          6              2    &quot;Korarchaeota&quot;    10      10           1
## 6          7              2   &quot;Nanoarchaeota&quot;     3       3           1
## 7          8              2 Nanohaloarchaeota     1       0           1
## 8          9              2  &quot;Thaumarchaeota&quot;     6       0           1
## 9         10              4      Thermoprotei    55       0           2
## 10        11             10      Acidilobales     3       0           3
## # ... with 1,989 more rows, and 5 more variables: n_subtaxa &lt;dbl&gt;,
## #   n_subtaxa_1 &lt;dbl&gt;, hierarchies &lt;chr&gt;, count_amplified &lt;dbl&gt;,
## #   prop_amplified &lt;dbl&gt;</code></pre>
<p>The proportion of sequences amplified can be represented by color in a metadiversity plot:</p>
<pre class="r"><code>set.seed(3)
heat_tree(pcr, node_size = n_obs, node_label = name,
          node_color = prop_amplified,
          node_color_range =  c(&quot;red&quot;, &quot;orange&quot;, &quot;yellow&quot;, &quot;green&quot;, &quot;cyan&quot;),
          node_color_trans = &quot;linear&quot;, tree_label = name)</code></pre>
<p><img src="index_files/figure-html/home_plot_8-1.png" width="672" /></p>
<p>This plot makes it apparent that most taxa were amplified, but not all. The data can also be subset to better see what did not get amplified:</p>
<pre class="r"><code>set.seed(1) 
pcr %&gt;%
  filter_taxa(count_amplified &lt; n_obs) %&gt;% 
  heat_tree(node_size = n_obs, node_label = name, node_color = prop_amplified,
            node_color_range =  c(&quot;red&quot;, &quot;orange&quot;, &quot;yellow&quot;, &quot;green&quot;, &quot;cyan&quot;),
            initial_layout = &quot;reingold-tilford&quot;, layout = &quot;davidson-harel&quot;,
            node_color_interval = c(0, 1), node_color_trans = &quot;linear&quot;,
            node_label_max = 500, tree_label = name)</code></pre>
<p><img src="index_files/figure-html/home_plot_9-1.png" width="672" /></p>
</div>
<div id="differential-heat-trees" class="section level2">
<h2>Differential heat trees</h2>
<p>We can use what we call <strong>differential heat trees</strong> to compare the values of treatments such as:</p>
<ul>
<li>the relative abundance of taxa in two communities</li>
<li>the coverages of different primer pairs</li>
</ul>
<p>Here, we compare the effectiveness 515F/806R to another primer pair 357F/519F, by plotting the difference in proportions amplified by each. First, the same sequences are amplified with 357F/519F and results for the two primer pairs combined:</p>
<pre class="r"><code>pcr_2 &lt;- primersearch(sampled,
                      forward = c(&quot;357F&quot; = &quot;CTCCTACGGGAGGCAGCAG&quot;), 
                      reverse = c(&quot;519F&quot; = &quot;GWATTACCGCGGCKGCTG&quot;), 
                      mismatch = 10)
pcr &lt;- mutate_taxa(pcr, 
                   count_amplified_2 = taxon_data(pcr_2, col_subset = &quot;count_amplified&quot;, drop = TRUE),
                   prop_diff = prop_amplified - taxon_data(pcr_2, col_subset = &quot;prop_amplified&quot;, drop = TRUE))</code></pre>
<p>Then, taxa that are not amplified by both pairs can be subset and the difference in amplification plotted. In the plot below, green corresponds to taxa amplified by 515F/806R but not 357F/519F and brown is the opposite:</p>
<pre class="r"><code>set.seed(2)
pcr %&gt;%
  filter_taxa(abs(prop_diff) &gt; 0.1, supertaxa = TRUE) %&gt;%
  heat_tree(node_size = n_obs,
            node_label = ifelse(abs(prop_diff) &gt; 0.3, name, NA),
            node_color = prop_diff,
            node_color_range = diverging_palette(),
            node_color_interval = c(-1, 1),
            node_color_trans = &quot;linear&quot;,
            node_label_max = 500,
            tree_label = name,
            node_color_axis_label = &quot;Difference in proportion amplified&quot;,
            node_size_axis_label = &quot;Sequence count&quot;,
            initial_layout = &quot;reingold-tilford&quot;, layout = &quot;davidson-harel&quot;,
            title = &quot;Comparison of two universal primer pairs using in silico PCR&quot;,
            title_size = 0.03)</code></pre>
<p><img src="index_files/figure-html/home_plot_10-1.png" width="672" /></p>
</div>
<div id="future-development" class="section level2">
<h2>Future development</h2>
<p>MetacodeR is under active development and many new features are planned. Some improvements that are being worked on include:</p>
<ul>
<li>Increases in function speed</li>
<li>Plotting functions for pairwise comparison of treatments</li>
<li>Barcoding gap analysis and associated plotting functions</li>
<li>A function to aid in retrieving appropriate sequence data from NCBI for <em>in silico</em> PCR from whole genome sequences.</li>
</ul>
<p>To see the details of what is being worked on, check out the <a href="https://github.com/grunwaldlab/metacoder/issues">issues</a> tab of the MetacodeR <a href="https://github.com/grunwaldlab">Github site</a>.</p>
</div>
<div id="feedback" class="section level2">
<h2>Feedback</h2>
<p>We would like to hear about users’ thoughts on the package and any errors they run into. Please report bugs and comments on the <a href="https://github.com/grunwaldlab/metacoder/issues">issues</a> tab of the MetacodeR <a href="https://github.com/grunwaldlab">Github site</a>. We also welcome contributions via a Github <a href="https://help.github.com/articles/using-pull-requests/">pull request</a>.</p>
</div>
<div id="aknowledgements" class="section level2">
<h2>Aknowledgements</h2>
<p>We thank Tom Sharpton for sharing his metagenomics expertise and advising us. MetacodeR’s major dependencies are taxize, igraph, and ggplot2.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p>[1] S. A. Olson. “Emboss opens up sequence analysis”. In: <em>Briefings in Bioinformatics</em> 3.1 (Jan. 2002), pp. 87-91. DOI: 10.1093/bib/3.1.87. &lt;URL: <a href="http://dx.doi.org/10.1093/bib/3.1.87" class="uri">http://dx.doi.org/10.1093/bib/3.1.87</a>&gt;.</p>
<p>[2] W. Walters, E. R. Hyde, et al. “Improved Bacterial 16S rRNA Gene (V4 and V4-5) and Fungal Internal Transcribed Spacer Marker Gene Primers for Microbial Community Surveys”. In: <em>mSystems</em> 1.1 (Dec. 2015). Ed. by H. Bik, pp. e00009-15. DOI: 10.1128/msystems.00009-15. &lt;URL: <a href="http://dx.doi.org/10.1128/mSystems.00009-15" class="uri">http://dx.doi.org/10.1128/mSystems.00009-15</a>&gt;.</p>
</div>

<div id="disqus_thread"></div>
<script>

/**
 *  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
 *  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables */
/*
var disqus_config = function () {
    this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
    this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
    var d = document, s = d.createElement('script');
    s.src = '//metacoder-documentation.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>

</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
