---
title: "Human microbiome example"
---

```{r greengenes_setup, echo=FALSE, warning=FALSE, message=FALSE}
source("settings.R")
```

## Requirements 

**NOTE:** This analysis requires at least 10Gb of RAM to run. 

## Parsing taxonomic data

```{r hmp_parse, warning=FALSE, message=FALSE}
library(metacoder)
v35_data <- parse_hmp_qiime(otu_file = "http://downloads.hmpdacc.org/data/HMQCP/otu_table_psn_v35.txt.gz",
                            mapping_file = "http://downloads.hmpdacc.org/data/HMQCP/v35_map_uniquebyPSN.txt.bz2")
```


## Plot of everything

```{r hmp_plot_all}
calculate_abundance <- function(data) {
  mutate_taxa(data, abundance = vapply(obs(data), function(i) sum(data$obs_data[i, colnames(data$obs_data) %in% data$mapping$sample_id]), numeric(1)))
}

v35_data <- calculate_abundance(v35_data)


plot_all <- function(data, output_name, seed = 1) {
  set.seed(seed)
  data %>%
    filter_taxa(abundance >= 100) %>%
    plot(node_size = n_obs,
         node_size_axis_label = "Number of OTUs",
         node_color = abundance,
         node_color_trans = "area",
         node_color_axis_label = "Number of reads",
         node_label = name,
         node_label_max = 100,
         overlap_avoidance = 1,
         # initial_layout = "re", layout = "da",
         output_file = file.path(output_folder, paste0(output_name, output_format)))
}

plot_all(v35_data, "hmp--v35--all_data")
```




## Plot body site differences

```{r hmp_plot_diff}
calculate_prop <- function(data, site) {
  sample_cols <- as.character(unlist(data$mapping[data$mapping$body_site == site, "sample_id"]))
  sample_cols <- sample_cols[sample_cols %in% colnames(data$obs_data)]
  obs_indexes <- obs(data)
  total_counts <- vapply(sample_cols, function(s) sum(data$obs_data[, s]), numeric(1))
  col_name <- paste0(site, "_median_prop")
  lapply(obs_indexes,
         function(i) {
           vapply(sample_cols, 
                  function(s) sum(data$obs_data[i, s]) / total_counts[s],
                  numeric(1))})
}

calculate_prop_diff <- function(data, site_1, site_2) {
  props_1 <- calculate_prop(data, site_1)
  props_2 <- calculate_prop(data, site_2)
  p_value <- mapply(function(x, y) wilcox.test(x, y)$p.value,
                    props_1, props_2)
  p_value <- p.adjust(p_value, method = "bonferroni")
  ifelse(p_value > 0.05 | is.nan(p_value), 0, vapply(props_1, median, numeric(1)) - vapply(props_2, median, numeric(1)))
}

plot_body_site_diff <- function(data, site_1, site_2, output_name, seed = 1) {
  # calculate site difference
  data$taxon_data$median_prop_diff <- calculate_prop_diff(data, site_1, site_2)
  
  # plot
  set.seed(seed)
  data %>%
    filter_taxa(abundance >= 100) %>%
    # filter_taxa(n_supertaxa <= 4) %>%
    plot(node_size_axis_label = "Number of OTUs",
         node_size = n_obs,
         node_color_axis_label = paste0(site_1, " only    Both    ", site_2, " only"),
         node_color = median_prop_diff,
         node_color_range = diverging_palette(),
         node_color_trans = "area",
         node_color_interval = max(abs(range(median_prop_diff))) * c(-1, 1),
         edge_color_interval = max(abs(range(median_prop_diff))) * c(-1, 1),
         node_label = name,
         node_label_max = 150,
         overlap_avoidance = 1,
         # initial_layout = "re", layout = "da",
         # maxiter = 20, fineiter = 20,
         # margin_size = c(0.001, 0.001),
         output_file = file.path(output_folder, paste0(output_name, "--", site_1, "_vs_", site_2, output_format)))
}

plot_body_site_diff(v35_data, "Throat", "Saliva", "hmp--v35")
plot_body_site_diff(v35_data,  "Supragingival_plaque", "Subgingival_plaque", "hmp--v35")
plot_body_site_diff(v35_data,  "Anterior_nares", "Buccal_mucosa", "hmp--v35")
```


## Plot body site difference matrix

```{r hmp_diff_matrix, fig.width = 7.5, fig.height = 7.5}
# Identify pairs of treatments to compare
body_sites <- c("Stool", "Saliva", "Tongue_dorsum", "Buccal_mucosa", "Anterior_nares")
combinations <- t(combn(seq_along(body_sites), 2))

# Make reduced taxonomy for plotting
v35_reduced <- v35_data
v35_reduced <- filter_taxa(v35_reduced, n_supertaxa < 4)

# Calculate significant body site differneces
apply(combinations, MARGIN = 1, function(i) {
  v35_reduced$taxon_data[[paste0(body_sites[i], collapse = "-")]] <<- calculate_prop_diff(v35_reduced, body_sites[i[1]],  body_sites[i[2]])
})
site_diff_cols <- apply(combinations, MARGIN = 1, function(i) paste0(body_sites[i], collapse = "-"))

# Remove data not needed for plotting
is_diff <- apply(v35_reduced$taxon_data[ , site_diff_cols], MARGIN = 1, function(a_row) any(a_row != 0))
v35_reduced <- filter_taxa(v35_reduced, is_diff, supertaxa = TRUE) # remove taxa with no differences
v35_reduced$obs_data <- v35_reduced$obs_data[, c("obs_taxon_ids", "otu_id")] # remove per-otu data since it is big

# Make individual plots
plot_body_site_diff_small <- function(data, diff_col, seed = 1) {
  set.seed(seed)
  mutate_taxa(data, diff_col = data$taxon_data[[diff_col]]) %>%
    # filter_taxa(is_diff, supertaxa = TRUE) %>%
    plot(node_size = n_obs,
         node_color = diff_col,
         node_color_range = diverging_palette(),
         node_color_trans = "linear",
         node_color_interval = c(-0.5, 0.5),
         edge_color_interval = c(-0.5, 0.5),
         # node_label = ifelse(abs(diff_col) > 0.01, name, NA),
         # node_label_max = 50,
         overlap_avoidance = 0.5,
         node_size_axis_label = "Number of OTUs",
         node_color_axis_label = "Median proportion difference",
         make_legend = FALSE)
}


sub_plots <- lapply(site_diff_cols, function(x) plot_body_site_diff_small(v35_reduced, x))
key_plot <- plot(v35_reduced,
                 node_size = n_obs,
                 node_color = "#DDDDDD",
                 node_color_range = diverging_palette(),
                 node_color_trans = "linear",
                 node_color_interval = c(-0.5, 0.5),
                 edge_color_interval = c(-0.5, 0.5),
                 node_label = ifelse(n_obs > 2000, name, NA),
                 node_label_max = 500,
                 edge_label = ifelse(n_obs > 2000, NA, name),
                 edge_label_max = 500,
                 edge_label_size_range = c(0.01, 0.02),
                 overlap_avoidance = 0.5,
                 node_size_axis_label = "Number of OTUs",
                 node_color_axis_label = "Median proportion difference")

matrix_plot <- ggdraw(plot_grid(plotlist = plots[layout_matrix], label_size = 8,
                                nrow = nrow(layout_matrix), 
                                ncol = ncol(layout_matrix), 
                                labels = site_diff_cols[layout_matrix])) + 
  draw_plot(key_plot, x = 0.5, y = 0, width = 0.5, height = 0.5)

save_plot(file.path(output_folder, "figure_3--hmp_matrix_plot.pdf"), 
          matrix_plot, base_width = 7.5, base_height = 7.5)
print(matrix_plot)
```